enclave/core/sgx/longjmp.S

```diff
@@ -22,16 +22,22 @@ oe_longjmp:
     // if (val == 0) then set val to 1.
     cmp $0, %rax
     jne val_is_nonzero
-    mov $1, %rax
+    mov $1, %rax # Return value (int)
 val_is_nonzero:
-    mov (%rdi), %rbx
-    mov 8(%rdi), %rbp
-    mov 16(%rdi), %r12
-    mov 24(%rdi), %r13
-    mov 32(%rdi), %r14
-    mov 40(%rdi), %r15
-    mov 48(%rdi), %rdx
-    mov %rdx, %rsp
-    mov 56(%rdi), %rdx
-    jmp *%rdx
+    # Some versions of clang generate code that relies on the return address
```

Comment:
> Some versions of clang generate code that relies on the return address of this stack frame to be the address that was actually jumped to.  Set the stored %rip value as the return address for the stack frame directly above the one we're returning to.

Reading the code, we are not updating the `return address of this stack frame to the address that was actually jumped to`. There is no write to the memory locations addressed by %rsp or %rbp.

Reply:
## Where are we setting the return address

`movq 16(%rdi),%rcx` retrieves the value of `%rip` stored in `env`. This is the return address that we're going to jump back to.

`movq (%rdi),%rdx` gets the stack pointer of the stack frame we're jumping back to

`movq %rcx,-0x8(%rdx)` sets the return address as the start of the stack frame directly above the stack frame we're returning to.

## As to your question about the code that necessitates this change

Here is roughly some code that clang-9 generates for `oe_ocall`. I have cut out some unnecessary stuff and tried to add comments to explain whats going on:
```
callq  135b70 <oe_setjmp> # setjmp will make longjmp return back to the following instruction
mov    -0x8(%rsp),%rcx # Get the first 8 bytes of the stack frame above this one (aka, the return address that was pushed onto the stack by the call instruction)
... {later on}
lea    -0x13(%rip),%rsi # get the address of the above mov instruction
cmp    %rsi,%rcx # check that the instruction we just retrieved is the same as the return address
mov    -0x40(%rbp),%rcx # Move 0xffffffffffffffff to rcx
cmovne %rcx,%rdx # if rcx and rdx are not equal, move -1 into %rdx
... {later on}
or     %rdx,%rsp # or rdx with rsp. If rdx was set to -1 earlier, rsp will now be a garbage value
pop    %rbp # pop causes rsp to be read from. If it is a garbage value, we will segfault.
retq
```

| Stack |
| ------ |
|     Frame 1   |
| {return address} |

Reply:
Basically we're tricking the calling function (oe_ocall) into thinking that it has arrived at the `mov    -0x8(%rsp),%rcx` instruction by returning from oe_setjmp, when in reality it has arrived there via a call to `oe_longjmp`.

Reply:
Found a cleaner way to fix the problem. Resolving this conversation.